runtime/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonInput.kt

```diff
@@ -142,15 +149,47 @@ internal class StreamingJsonInput internal constructor(
         }
     }
 
-    override fun decodeBoolean(): Boolean = reader.takeString().run { if (configuration.strictMode) toBooleanStrict() else toBoolean() }
-    override fun decodeByte(): Byte = reader.takeString().toByte()
-    override fun decodeShort(): Short = reader.takeString().toShort()
-    override fun decodeInt(): Int = reader.takeString().toInt()
-    override fun decodeLong(): Long = reader.takeString().toLong()
-    override fun decodeFloat(): Float = reader.takeString().toFloat()
-    override fun decodeDouble(): Double = reader.takeString().toDouble()
-    override fun decodeChar(): Char = reader.takeString().single()
-    override fun decodeString(): String = reader.takeString()
-    override fun decodeEnum(enumDescriptor: SerialDescriptor): Int =
-        enumDescriptor.getElementIndexOrThrow(reader.takeString())
+    override fun decodeBoolean(): Boolean {
+        /*
+         * We prohibit non true/false boolean literals at all as it is considered way too error-prone,
+         * but allow quoted literal in relaxed mode for booleans.
+         */
+        return if (configuration.isLenient) {
+            reader.takeString().toBooleanStrict()
+        } else {
+            reader.takeBooleanStringUnquoted().toBooleanStrict()
+        }
+    }
+
+    /*
+     * The rest of the primitives are allowed to be quoted and unqouted
+     * to simplify integrations with third-party API.
+     */
+    override fun decodeByte(): Byte = reader.takeString().parse("byte") { toByte() }
+    override fun decodeShort(): Short = reader.takeString().parse("short") { toShort() }
+    override fun decodeInt(): Int = reader.takeString().parse("int") { toInt() }
+    override fun decodeLong(): Long = reader.takeString().parse("long") { toLong() }
+    override fun decodeFloat(): Float = reader.takeString().parse("float") { toFloat() }
+    override fun decodeDouble(): Double = reader.takeString().parse("double") { toDouble() }
+    override fun decodeChar(): Char = reader.takeString().parse("char") { single() }
+
+    override fun decodeString(): String {
+        return if (configuration.isLenient) {
+            reader.takeString()
+        } else {
+            reader.takeStringQuoted()
+        }
+    }
+
+    private inline fun <T> String.parse(type: String, block: String.() -> T): T {
+        try {
+            return block()
+        } catch (e: Throwable) {
+            reader.fail("Failed to parse $type")
```

Comment:
It would be nice to leave 'causedBy' on platforms where it is supported (JVM).

Reply:
It is generally a good point, but in this particular case (and in its `Tree` counterpart) cause will just duplicate the stack trace and won't really add any meaningful information.
The name of the primitive type and its values as part of the JSON input string already covers it